src/main/java/uk/gov/ons/ctp/response/casesvc/message/CaseReceiptReceiver.java

```diff
@@ -1,15 +1,81 @@
 package uk.gov.ons.ctp.response.casesvc.message;
 
+import static uk.gov.ons.ctp.response.casesvc.representation.CategoryDTO.CategoryName.OFFLINE_RESPONSE_PROCESSED;
+import static uk.gov.ons.ctp.response.casesvc.representation.CategoryDTO.CategoryName.ONLINE_QUESTIONNAIRE_RESPONSE;
+import static uk.gov.ons.ctp.response.casesvc.representation.CategoryDTO.CategoryName.PAPER_QUESTIONNAIRE_RESPONSE;
+import static uk.gov.ons.ctp.response.casesvc.utility.Constants.QUESTIONNAIRE_RESPONSE;
+import static uk.gov.ons.ctp.response.casesvc.utility.Constants.SYSTEM;
+
+import com.godaddy.logging.Logger;
+import com.godaddy.logging.LoggerFactory;
+import java.sql.Timestamp;
+import java.util.UUID;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.integration.annotation.MessageEndpoint;
+import org.springframework.integration.annotation.ServiceActivator;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
 import uk.gov.ons.ctp.common.error.CTPException;
+import uk.gov.ons.ctp.response.casesvc.domain.model.Case;
+import uk.gov.ons.ctp.response.casesvc.domain.model.CaseEvent;
 import uk.gov.ons.ctp.response.casesvc.message.feedback.CaseReceipt;
+import uk.gov.ons.ctp.response.casesvc.message.feedback.InboundChannel;
+import uk.gov.ons.ctp.response.casesvc.representation.CategoryDTO;
+import uk.gov.ons.ctp.response.casesvc.service.CaseService;
 
 /** The reader of CaseReceipts from queue */
-public interface CaseReceiptReceiver {
+@MessageEndpoint
+public class CaseReceiptReceiver {
+  private static final Logger log = LoggerFactory.getLogger(CaseReceiptReceiver.class);
+
+  private static final String EXISTING_CASE_NOT_FOUND = "No existing case found";
+
+  @Autowired private CaseService caseService;
+
   /**
    * To process CaseReceipts read from queue
    *
    * @param caseReceipt to process
-   * @throws CTPException when case state transition error
+   * @throws CTPException CTPException
    */
-  void process(CaseReceipt caseReceipt) throws CTPException;
+  @Transactional(propagation = Propagation.REQUIRED, readOnly = false, value = "transactionManager")
+  @ServiceActivator(inputChannel = "caseReceiptTransformed", adviceChain = "caseReceiptRetryAdvice")
+  public void process(CaseReceipt caseReceipt) throws CTPException {
```

Comment:
not sure if the contract in this signature is correct . At the end you are looking for 
1. inboundChannel
2. caseId 
that you retrieve via the caseRecepit. 

I would have encapusapte this outside possibly and move the transactional at lower level in this method.

Reply:
We could do some refactoring of this as you suggest, but we've simply gotten rid of the Impl classes which were unnecessary. The code has been copied from CaseReceiptReceiverImpl.java and is unchanged. The 'contract' is correct because it's a Rabbit message receiver method receiving an object of that type.